libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/events/SchemaTest.kt

```diff
@@ -0,0 +1,278 @@
+                }
+            }
+            if (field.equals("distanceRemaining", ignoreCase = true)) {
+                distanceRemainingCount++
+                if (distanceRemainingCount > 1) {
+                    field = "step$field"
+                }
+            }
+            val thisSchema = findSchema(properties, field)
+            Assert.assertNotNull(field, thisSchema)
+            if (thisSchema.has("type")) {
+                typesMatch(thisSchema, type)
+            }
+        }
+    }
+
+    private fun findSchema(schema: JsonObject, field: String): JsonObject {
+        return schema.getAsJsonObject(field)
+    }
+
+    private fun typesMatch(schema: JsonObject, type: String) {
+        val eventType = if (type.equals("int", ignoreCase = true) ||
+            type.equals("integer", ignoreCase = true) ||
+            type.equals("double", ignoreCase = true) ||
+            type.equals("float", ignoreCase = true)
+        ) {
+            "number"
+        } else if (type.contains("[]")) {
+            "array"
+        } else if (type.equals("string", ignoreCase = true)) {
+            "string"
+        } else if (type.equals("AppMetadata", ignoreCase = true)) {
+            "object"
+        } else {
+            type
+        }
+        val typeClass: Class<out JsonElement> = schema["type"].javaClass
+        val jsonElement = JsonParser().parse(eventType.toLowerCase())
+        if (typeClass == JsonPrimitive::class.java) {
+            val typePrimitive = schema["type"]
+            Assert.assertEquals(typePrimitive, jsonElement)
+        } else {
+            val arrayOfTypes = schema.getAsJsonArray("type")
+            Assert.assertTrue(arrayOfTypes.contains(jsonElement))
+        }
+    }
+
+    private fun grabSchemaPropertyFields(aClass: Class<*>): List<Field> {
+        val fields: MutableList<Field> = ArrayList()
+        val allFields = aClass.declaredFields
+        for (field in allFields) {
+            if (field.type == NavigationStepData::class.java) {
+                val dataFields =
+                    field.type.declaredFields
+                for (dataField in dataFields) {
+                    if (Modifier.isPrivate(dataField.modifiers) && !Modifier.isStatic(dataField.modifiers)) {
+                        fields.add(dataField)
+                    }
+                }
+            } else if (Modifier.isPrivate(field.modifiers) && !Modifier.isStatic(field.modifiers)) {
+                fields.add(field)
+            }
+        }
+        val superFields = aClass.superclass!!.declaredFields
+        for (field in superFields) {
+            if (Modifier.isPrivate(field.modifiers) && !Modifier.isStatic(field.modifiers)) {
+                fields.add(field)
+            }
+        }
+
+        // filter out non-property fields
+        fields.removeIf { it.name == "version" }
+        return fields
+    }
+
+    private fun grabEventSchema(eventName: String, version: String = "2.0"): EventSchema =
```

Comment:
Why aren't we defaulting to `2.1`? 🤔 I think that's the answer to 👇 https://github.com/mapbox/mapbox-navigation-android/pull/3339#pullrequestreview-448493579 right?

> Wondering why legacy `SchemaTest` is working and CI ✅ as this part is changing only `SchemaTest` from `libnavigation-core` but not the one from `libnavigation-android` 🤔

Reply:
The rest of the navigation events are `v2.0`, only the feedback event has been updated.

Reply:
And yes, that's why the legacy tests pass, I've only updated the 1.0 events to use the `v2.1` feedback event schema.

Reply:
Curious, is the event `version` something new or something we weren't relying on before?

Reply:
The backend has a default fallback for all events to version `2.0`. Since all our events were `v2.0` we did not have to specify this in the past. Now, with a new version of the feedback schema, we need to pass it in the event as well. Whenever we do refactor the telemetry classes, we might want to look into how the fields are structured to better match the schema.

Reply:
> The backend has a default fallback for all events to version `2.0`. Since all our events were `v2.0` we did not have to specify this in the past. Now, with a new version of the feedback schema, we need to pass it in the event as well.

That makes total sense and what I was imagining. Historically, all navigation events have had the same version but not anymore as you're describing. Now it seems that every event has its own versioning. Thanks for clarifying 👍 

> Whenever we do refactor the telemetry classes, we might want to look into how the fields are structured to better match the schema.

Totally agree. Also the `SchemaTest` is not checking for correctness is only checking the event type fields (not checking the schema rules).